### PR TITLE
Move diplomacy to standalone library absorbing aop

### PR DESCRIPTION
**Pull request by lordspacehog**
**Related issue**:  https://github.com/chipsalliance/rocket-chip/issues/3037

**Type of change**: other enhancement

**Impact**: API Deprecation in favor of standalone diplomacy

**Development Phase**: implementation

**Release Notes**
Splits out the aop and diplomacy modules into a standalone library. Adds the new library as a dependency and updates the in-tree diplomacy and aop modules to reference the new library with deprecation warnings.

### CODE REVIEW

**Comment by jerryz123** [comment]
Very nice, I greatly appreciate the effort to preserve the deprecated package paths.

**Comment by lordspacehog** [comment]
yeah, I mostly followed the lead of @sequencer and attempted to capture as minimal of a change as possible for this initial split. I also have other PRs ready to go that update each of the individual modules internally to use the new diplomacy library directly.

I've also run tests locally and everything seems to be passing without issue. so this should be ready for review.

File: .scalafmt.conf [hunk 1]
```diff
@@ -0,0 +1,32 @@
+version = 3.7.17
```
Comment (sequencer) [inline .scalafmt.conf:1]: Don't add Scalafmt in to RocketChip in this commit(although I also like to do it)

File: .mill-version [hunk 2]
```diff
@@ -0,0 +1 @@
+0.11.7
```
Comment (sequencer) [inline .mill-version:1]: This file is not required, the rocket-chip use nix to handle the version. it will introduce downloading artifacts from the internet.

File: .gitmodules [hunk 3]
```diff
@@ -7,3 +7,6 @@
 [submodule "dependencies/chisel"]
 	path = dependencies/chisel
 	url = https://github.com/chipsalliance/chisel.git
+[submodule "dependencies/diplomacy"]
+	path = dependencies/diplomacy
+	url = https://github.com/lordspacehog/diplomacy.git
```
Comment (sequencer) [inline .gitmodules:12]: should be chipsalliance/diplomacy in the future.

File: build.sc [hunk 4]
```diff
@@ -4,41 +4,39 @@ import mill.scalalib.publish._
...
 object v {
-  val scala = "2.13.12"
+  val scala               = "2.13.12"
   // the first version in this Map is the mainly supported version which will be used to run tests
   val chiselCrossVersions = Map(
-    "5.1.0" -> (ivy"org.chipsalliance::chisel:5.1.0", ivy"org.chipsalliance:::chisel-plugin:5.1.0"),
+    "5.1.0" -> (ivy"org.chipsalliance::chisel:5.1.0", ivy"org.chipsalliance:::chisel-plugin:5.1.0")
     // build from project from source
-    "source" -> (ivy"org.chipsalliance::chisel:99", ivy"org.chipsalliance:::chisel-plugin:99"),
+    // "source" -> (ivy"org.chipsalliance::chisel:99", ivy"org.chipsalliance:::chisel-plugin:99")
```
Comment (sequencer) [inline build.sc:17]: unrelated change.

File: build.sc [hunk 5]
```diff
@@ -4,41 +4,39 @@ import mill.scalalib.publish._
...
 
-trait Chisel
-  extends millbuild.dependencies.chisel.build.Chisel {
-  def crossValue = v.scala
+trait Chisel extends millbuild.dependencies.chisel.build.Chisel {
+  def crossValue              = v.scala
   override def millSourcePath = os.pwd / "dependencies" / "chisel"
-  def scalaVersion = T(v.scala)
+  def scalaVersion            = T(v.scala)
 }
```
Comment (sequencer) [inline build.sc:35]: nit: don't format it to keep the patch clean.

File: build.sc [hunk 6]
```diff
@@ -4,41 +4,39 @@ import mill.scalalib.publish._
...
 
 object macros extends Macros
 
-trait Macros
-  extends millbuild.common.MacrosModule
-    with RocketChipPublishModule
-    with SbtModule {
+trait Macros extends millbuild.common.MacrosModule with RocketChipPublishModule with SbtModule {
```
Comment (sequencer) [inline build.sc:39]: I think Macros can be removed, since the `ValName` macro is from diplomacy now.

File: build.sc [hunk 7]
```diff
@@ -48,7 +46,7 @@ trait Macros
 object hardfloat extends mill.define.Cross[Hardfloat](v.chiselCrossVersions.keys.toSeq)
 
 trait Hardfloat
-  extends millbuild.dependencies.hardfloat.common.HardfloatModule
+    extends millbuild.dependencies.hardfloat.common.HardfloatModule
```
Comment (sequencer) [inline build.sc:49]: nit: don't format it to keep the patch clean.

File: build.sc [hunk 8]
```diff
@@ -69,20 +67,45 @@ trait Hardfloat
...
-trait CDE
-  extends millbuild.dependencies.cde.common.CDEModule
-    with RocketChipPublishModule
-    with ScalaModule {
+trait CDE extends millbuild.dependencies.cde.common.CDEModule with RocketChipPublishModule with ScalaModule {
 
   def scalaVersion: T[String] = T(v.scala)
 
   override def millSourcePath = os.pwd / "dependencies" / "cde" / "cde"
 }
```
Comment (sequencer) [inline build.sc:75]: nit: don't format it to keep the patch clean.

File: build.sc [hunk 9]
```diff
@@ -4,41 +4,39 @@ import mill.scalalib.publish._
...
   )
-  val mainargs = ivy"com.lihaoyi::mainargs:0.5.0"
-  val json4sJackson = ivy"org.json4s::json4s-jackson:4.0.5"
-  val scalaReflect = ivy"org.scala-lang:scala-reflect:${scala}"
-  val sonatypesSnapshots = Seq(
+  val mainargs            = ivy"com.lihaoyi::mainargs:0.5.0"
+  val json4sJackson       = ivy"org.json4s::json4s-jackson:4.0.5"
+  val scalaReflect        = ivy"org.scala-lang:scala-reflect:${scala}"
+  val sonatypesSnapshots  = Seq(
     MavenRepository("https://s01.oss.sonatype.org/content/repositories/snapshots")
```
Comment (sequencer) [inline build.sc:23]: nit: don't format it to keep the patch clean.

File: build.sc [hunk 10]
```diff
@@ -69,20 +67,45 @@ trait Hardfloat
...
+}
+
 object rocketchip extends Cross[RocketChip](v.chiselCrossVersions.keys.toSeq)
 
 trait RocketChip
-  extends millbuild.common.RocketChipModule
+    extends millbuild.common.RocketChipModule
     with RocketChipPublishModule
     with SbtModule
     with Cross.Module[String] {
```
Comment (sequencer) [inline build.sc:111]: nit: don't format it to keep the patch clean.

File: build.sc [hunk 11]
```diff
@@ -104,15 +127,20 @@ trait RocketChip
...
+  def diplomacyModule = diplomacy(crossValue)
+
+  def diplomacyIvy = None
+
   def mainargsIvy = v.mainargs
 
   def json4sJacksonIvy = v.json4sJackson
 
+  def sourcecodeIvy = v.sourcecode
+
```
Comment (sequencer) [inline build.sc:139]: `soucecodeIvy` is not required?

File: common.sc [hunk 12]
```diff
@@ -1,8 +1,7 @@
 import mill._
```
Comment (sequencer) [inline common.sc:1]: nit: don't format it to keep the patch clean.

File: src/main/scala/diplomacy/Resources.scala [hunk 13]
```diff
@@ -10,20 +10,26 @@ import scala.collection.mutable.HashMap
 sealed trait ResourceValue
```
Comment (sequencer) [inline src/main/scala/diplomacy/Resources.scala:10]: nit: don't format it to keep the patch clean.

**Review by sequencer — CHANGES_REQUESTED** [review]
Basically LTGM, really really really thank to your rebasing!!!
There are some nitpits in the commit. please keep the patch as minimal as possible. I'll take a look at your reabsed version of diplomacy today!

File: .gitmodules (hunk 3 above)
Comment (lordspacehog) [reply .gitmodules:12]: yeah, My current plan is to do a 2nd PR to cut it over,. or we can wait until diplomacy is published to ivy and i can just the setup over to use ivy.

**Comment by lordspacehog** [comment]
@sequencer This new diplomacy split is a new history pull from rocket-chip/dev using git filter-repo (https://github.com/newren/git-filter-repo). You should either force push over your local master or use a different branch name pointed to the new history for now. 

The exact process i used was as follows:
1) Fork sequencer/diplomacy
2) clean checkout rocket-chip/dev in new folder.
3) use git filter-repo to prune all history unrelated to the changes to the diplomacy module.
4) add my diplomacy fork as upstream and push new history to main branch
5) move master branch to master-archive and set main as the default branch

From there i manually went through and reconciled the changes between your diplomacy repo, the main chipsalliance diplomacy repo, and the changes that happened directly in the rocket-chip repo.

**Comment by sequencer** [comment]
@lordspacehog I sent u an invitation, please directly push to a branch to chipsalliance/diplomacy, and I can take a look ;p

**Comment by lordspacehog** [comment]
@sequencer accepted the invite, but it looks like i don't have push access to the repo, only triage. if you can update that i'll get a branch pushed ASAP. i'll also stage an update to this PR to set that new branch as the submodule target.

**Comment by sequencer** [comment]
> @sequencer accepted the invite, but it looks like i don't have push access to the repo, only triage. if you can update that i'll get a branch pushed ASAP. i'll also stage an update to this PR to set that new branch as the submodule target.

Access granted.

**Comment by lordspacehog** [comment]
@sequencer pushed, https://github.com/chipsalliance/diplomacy/tree/lordspacehog/standalone_diplomacy

File: .gitmodules (hunk 3 above)
Comment (lordspacehog) [reply .gitmodules]: updated this PR to point at the new branch in the chipsalliance/diplomacy repo

**Comment by sequencer** [comment]
cc @lordspacehog, I reviewed and pushed the master branch to https://github.com/chipsalliance/diplomacy. please set that as upstream and go though CI;p

**Comment by lordspacehog** [comment]
@sequencer updated the submodule reference.

**Review by sequencer — APPROVED** [review]

**Comment by sequencer** [comment]
it might have additional things to clean up. Merge it for now. I’ll take care of them later this week
